### PR TITLE
BLD: Try g0 on macos?

### LINKED ISSUE #51900: ENH: strip debug symbols from macOS wheels
### Feature Type

- [ ] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [X] Removing existing functionality in pandas

### Problem Description

I found tonight that `.so` files in `pandas` macOS wheels contain debug symbols.

By my estimate, these add around 1MB compressed and around 4 MB uncompressed to the `cp311-cp311-macosx_10_9_universal2` wheel.

<details><summary>how I estimated that (click me)</summary>

On my macbook tonight (macOS 12.2, Intel CPU), I downloaded the latest CPython 3.11, macOS universal wheel and checked its size with `du`.

```shell
WHEEL_FILENAME='pandas-1.5.3-cp311-cp311-macosx_10_9_universal2.whl'
du -h ${WHEEL_FILENAME}
# 18M
```

Next, I unzipped it and used `dsymutil` to check for debug symbols.

```shell
mkdir unpack-dir
cp \
    ./${WHEEL_FILENAME} \
    ./unpack-dir

cd unpack-dir
unzip ./${WHEEL_FILENAME}

# check uncompressed size
du -sh pandas
# 65M

du -h pandas/_libs/missing.cpython-311-darwin.so
# 524K

dsymutil -s pandas/_libs/missing.cpython-311-darwin.so | grep N_OSO
```

That showed some (and interesting that they seem to include filepaths from a CI system's runner image 👀 )

```text
# [   765] 00007670 66 (N_OSO        ) 03     0001   0000000063c8cc76 '/Users/runner/work/1/s/pandas/build/temp.macosx-10.9-x86_64-cpython-311/pandas/_libs/missing.o'
# [   261] 00002a95 66 (N_OSO        ) 00     0001   0000000063c8cef7 '/Users/runner/work/1/s/pandas/build/temp.macosx-11.0-arm64-cpython-311/pandas/_libs/missing.o'
```

Next, I tried stripping all of the `.so` files using OSX `strip` ([docs link](https://www.unix.com/man-page/osx/1/strip/)).

```shell
find \
    $(pwd)/pandas \
    -type f \
    -name '*.so' \
    -exec strip -S '{}' \;
```

> **Warning**
>
> This did produce warnings like the following: `/Library/Developer/CommandLineTools/usr/bin/strip: warning: changes being made to the file will invalidate the code signature in: /private/tmp/check-pandas/unpack-dir/pandas/_libs/interval.cpython-311-darwin.so (for architecture arm64)`
>
> So maybe that's not the best approach for `pandas` actual build pipeline.

Packed the wheel back up

```shell
# check uncompressed size again
du -sh pandas
# 61M

rm ./${WHEEL_FILENAME}
zip -r ${WHEEL_FILENAME} .

# check compressed size again
du -h ${WHEEL_FILENAME}
# 17M
```

Then installed it (making sure the environment didn't have `pandas` installed before), and ran the tests.

```shell
pip uninstall --yes pandas
pip install ${WHEEL_FILENAME}
pip install pytest hypothesis
python -c "import pandas as pd; pd.test()"
```

Installation worked, and the tests all ran to completion, with the following results:

```text
= 1 failed, 152571 passed, 24052 skipped, 1362 xfailed, 12 xpassed, 1762 warnings, 23 errors in 1144.98s (0:19:04) =
```

</details>

I checked all of the wheels (for all platforms) from the 1.5.3 release ([PyPI link](https://pypi.org/project/pandas/1.5.3/#history)) and only found debug symbols in the macOS ones.

I did not repeat the analysis above to estimate the size impact of those symbols for any wheels other than the `cp311-cp311-macosx_10_9_universal2` one.

### Feature Description

If the inclusion of these symbols is not intentional and if I'm right that they're not necessary, please consider removing them.

This might be accomplished by avoiding them in the first place, e.g.:

* preventing compiler flag `-g` or similar from being used at build time
    - e.g. by setting `-DCMAKE_BUILD_TYPE=Release` if `CMake` is being used)*
* passing build-time flags to the linker or compiler which explicitly strips debug symbols
    - e.g. `-Wl,strip-all` for `gcc` as suggested in https://github.com/pandas-dev/pandas/issues/19531#issuecomment-362914720

Or by stripping those built objects after the fact.

* e.g. with `strip -S _file.so` on MacOS

I'm not familiar enough with `pandas` build system and preferred toolchain to offer more specific recommendations, sorry.

### Alternative Solutions

Instead of a `pandas`-specific fix, it might be worth adding support to [`delocate`](https://github.com/matthew-brett/delocate) similar to how [`auditwheel`](https://github.com/pypa/auditwheel) supports stripping after the fact with `auditwheel repair --strip` for Linux.

I'm not aware of another such tool that works with macOS wheels containing mach-o format objects.

### Additional Context

**Relevant Discussions**

I can see there was some discussion about stripping this project's wheels back in 2018, although that looks to be mostly about Linux wheels:

* https://github.com/MacPython/pandas-wheels/issues/24
* https://github.com/multi-build/multibuild/pull/140
* https://github.com/pandas-dev/pandas/issues/19681#issuecomment-365377295
* https://github.com/pandas-dev/pandas/issues/19531
* https://github.com/pandas-dev/pandas/issues/20719

This conversation from 2020 contains some details about building `pandas` on macOS with `clang:

* https://github.com/pandas-dev/pandas/issues/33314

Other misc. discussions from similar projects about stripping debug symbols while building Python wheels.

* https://github.com/pypa/auditwheel/issues/63
* https://github.com/numpy/numpy/pull/16110
* https://github.com/pypa/cibuildwheel/issues/331
* https://github.com/MacPython/numpy-wheels/pull/82

**Notes for Reviewers**

Thanks very much for your time and consideration!

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #51900  (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by WillAyd** [comment]
Is the issue still seeing debug symbols on macOS? I think there is a Darwin option for `gused` that is the default which may be generating that:

https://gcc.gnu.org/onlinedocs/gcc/Darwin-Options.html

**Comment by lithomas1** [comment]
> Is the issue still seeing debug symbols on macOS? I think there is a Darwin option for `gused` that is the default which may be generating that:
> 
> https://gcc.gnu.org/onlinedocs/gcc/Darwin-Options.html

Doesn't Apple use clang nowadays?

In the logs, I'm seeing a ``-g`` option being passed to clang, so I'm trying to cancel it out with the -g0 option (I read somewhere on Stack Overflow that that works, will need to see results, though).

**Comment by lithomas1** [comment]
I'm not sure who's passing it though.

**Comment by WillAyd** [comment]
Yea `gcc` is an alias for clang on macOS by default. I think if you are using conda-compilers though it may actually install gcc.

AFAIK the APIs between the two are supposed to be similar, but I didn't see any documentation for what clang has on gused

https://clang.llvm.org/docs/ClangCommandLineReference.html

**Comment by WillAyd** [comment]
> In the logs, I'm seeing a `-g` option being passed to clang, so I'm trying to cancel it out with the -g0 option (I read somewhere on Stack Overflow that that works, will need to see results, though).

Yea I think that is true. The GCC documentation for options is really good and thorough. This also confirms that `-g0` should negate `-g`. Let's see

https://gcc.gnu.org/onlinedocs/gcc/Debugging-Options.html#Debugging-Options

**Comment by WillAyd** [comment]
Where in the logs do you see -g showing up? I must be looking in the wrong places. 

If I had to guess it's something conda might be adding for us. I think Ralf pointed that out on the meson PR, which was pretty enlightening.

**Comment by lithomas1** [comment]
> > Where in the logs do you see -g showing up? I must be looking in the wrong places.
> > If I had to guess it's something conda might be adding for us. I think Ralf pointed that out on the meson PR, which was pretty enlightening.
> 
> Wheels are built with cibuildwheel(uses the GHA python's and sometimes Docker on Linux), so there's absolutely no conda involved right now.
> 
> Right now what's really infuriating me is that, even though I'm passing -g0 via CFLAGS, it is somehow showing up as -g, and I'm not sure what I'm doing wrong.
> 
> ```
> gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -g -arch x86_64 -DNPY_NO_DEPRECATED_API=0 -I/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-build-env-e0sc4bd0/overlay/lib/python3.8/site-packages/numpy/core/include -I/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibw-run-am2shr8g/cp38-macosx_x86_64/build/venv/include -I/Library/Frameworks/Python.framework/Versions/3.8/include/python3.8 -c pandas/_libs/groupby.c -o build/temp.macosx-10.9-x86_64-cpython-38/pandas/_libs/groupby.o -Wno-error=deprecated-declarations -Wno-error=unreachable-code
> ```
> 
> Here is an example command. The second `-g`(before `-arch`) is mine, but the first one is from something(`sysconfig`? default flags).
> 
> You can find this by ctrl+f for gcc in the logs.

Well, I'm just being stupid. It turns out I was overriding the cibuildwheel environment vars for macos 🤦.

File: pyproject.toml [hunk 1]
```diff
@@ -166,7 +166,7 @@ test-command = ""
 # macOS doesn't support stripping wheels with linker
 # https://github.com/MacPython/numpy-wheels/pull/87#issuecomment-624878264
 select = "*-macosx*"
-environment = ""
+environment = {CFLAGS="-g0"}
```
Comment (WillAyd) [inline pyproject.toml:169]: nice find. a bit orthogonal but wonder if we should strip all symbols here too

Comment (WillAyd) [reply pyproject.toml:169]: Guess it would help if I read the comment two lines up

**Comment by lithomas1** [comment]
@jameslamb 

I think I got the macOS symbols as well now. Savings are on par with what you mentioned 

Savings are currently 0.4 MB compressed for the x86_64 wheel, and 0.2 MB for the arm64 wheel.
(Since universal2 has both x86_64 and arm64, that's why the savings here are like half your savings. We also don't make those wheels anymore as of 2.0rc0, since numpy dropped them as well).

If you'd like to check one of the wheels yourself, here is a link to one of the wheels
https://github.com/pandas-dev/pandas/suites/11567086673/artifacts/599127251
(contains both x86_64 and arm64 in that zip file)

**Review by WillAyd — APPROVED** [review]

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lithomas1

**Comment by jameslamb** [comment]
> @jameslamb

> I think I got the macOS symbols as well now. Savings are on par with what you mentioned...

Thanks so much again for the quick resolution on this! I just pulled those wheels from the link you provided... don't see debug symbols in either of them 🎉 

If it interests you... I originally found these debug symbols (#51900) using a side-project I've been working on since last year's SciPy conference: https://github.com/jameslamb/pydistcheck. It's intended to be a CLI you run against wheels and sdists in CI, to catch issues like this ([click here for a full list of what it checks](https://pydistcheck.readthedocs.io/en/latest/check-reference.html)) and ensure that changes to your build process don't introduce regressions like letting the debug symbols back in.

Anyway, cheers and enjoy the rest of your week 👋🏻

**Pull request merged**